### DISCUSSION TITLE
Strapi production deployments in kubernetes

### DISCUSSION THREAD

**Discussion by martaver** [strapi/strapi#7638] [Legacy Topics]
Hi guys, I'm trying to wrap my head around the 'lifecycle' of a strapi production server.

I'm implementing strapi as the content store for the organisation I'm working in. I'd like to deploy strapi in an autoscaling configuration in our company's kubernetes cluster. I wonder if anyone has had experience achieving this?

I want changes to content types and customisations to happen in the development environment and be deployed to production... I want to avoid a scenario where I have to manually make updates to content types in the production server.

I understand that I can use a plugin to ensure that uploads are stored in cloud storage buckets, and then since the application state and content is kept in postgres, it seems like that should work.

My plan is to disable the content type editor for production, bundle the production build into the base docker image and deploy this to a cluster. This should work.

My big question is - how can I handle subsequent updates to the content types schema? In develop mode we will add types as our front-ends change. The strapi app simultaneously updates the database and the code in the strapi... I can bundle the updated code and push it out in a docker image... but how can I sync the content types in the production environment's database?

**Comment by derrickmehaffy** [comment]
Hi @martaver 

By default the content-type builder is disabled in production (when the `NODE_ENV` is anything but `development` which is the default AND when the start command is not the `develop` one)

When creating or updating a content-type (settings saved to the settings file) Strapi will only create new tables/columns but will not delete or rename them within the database (see this issue for more information as well as the discussion: https://github.com/strapi/strapi/issues/1114 )

If you are editing these within the content-type builder (renaming a field) right now we are only handling that as a visual change in which it doesn't update the actual settings, this is stored in the `core_store` table as a visual change to offset the above linked issue for now. You can export/import these settings with the [config:dump](https://strapi.io/documentation/v3.x/cli/CLI.html#strapi-configuration-dump-config-dump) and [config:restore](https://strapi.io/documentation/v3.x/cli/CLI.html#strapi-configuration-restore-config-restore) commands.

We know there is some painful issue when it comes to automated migrations and it's something we hope to tackle in our new database refresh: https://portal.productboard.com/strapi/1-roadmap/c/73-database-layer-v4

**Reply by martaver** [reply]
Hi @derrickmehaffy thanks for your reply... I'm a bit confused about creating types between what you're calling the 'settings file', the code that's generated in develop mode vs the data in postgres... 

Is it possible to disable the content-type builder altogether and simply define a schema in code? Will that migrate an older db to a new one when I deploy the server?

**Reply by derrickmehaffy** [reply]
settings files for models are in `api/*/models/*.settings.json` those define the schema (these are generally generated by the CTB), for the moment no it's not possible to entirely disable the CTB as it's a core package, it's also not easily removed. Plugins like the CTM (content-type manager) are directly integrated into it for controlling the views.

For migrations Strapi only creates new tables/columns, as I mentioned before it will not rename old ones, and will not delete them from the database due to #1114

**Reply by martaver** [reply]
Okay, so the approach is... use the CTB in develop mode to get things structured the way I want them, then do a production build and deploy. Then, for subsequent updates, I can use CTB in develop mode to make changes, then do a production build and deploy - the updated production build will automatically migrate the schema (except for deletes and renames)? 

Also, will I need persistent volumes if I have postgres and I'm using cloud storage for uploads?

**Reply by derrickmehaffy** [reply]
Exactly @martaver (for the schema), for renames/deletes you will need to handle this migration "manually". When I say manually here, you will need to find the best solution for you or your team. Generally what I do on personal projects is test it in staging and create a SQL migration script (again testing this script in staging as well) and once I'm happy with the result, this gets added into my repo to be executed by my Ansible deployment script (taking DB backups of course).

We do plan to address the migration process for renaming and cleaning up the database, we just need to ensure that novice users don't go blowing away a ton of their content without a database backup :sweat_smile: 

You should not need a persistent storage in production if you are using cloud uploads and a remote database, the entirety of your persistent code should be in your source control and using environment variables where needed.

**Reply by martaver** [reply]
Phew, that's a relief! As a strapi newbie, it was a little bit hard to tell with all the code updates that develop mode was making!

Thanks for the clarifications, they really helped!

**Comment by ambujsoni** [comment]
Hope your query is solved. However, I have published a complete guide regarding deploying [Strapi on production using Kubernetes (GKE), MySQL, and AWS S3.](https://ambujsoni.medium.com/deploy-strapi-on-production-using-kubernetes-gke-mysql-and-aws-s3-part-1-fec7ce6092a0)
Perhaps this could be of some help.

**Reply by derrickmehaffy** [reply]
@malgamves
